### PR TITLE
Serve front-end in `PrivateServer` mode

### PR DESCRIPTION
**Pull request by calyptobai**
This adds a new option `--frontend-dist` which accepts a path to serve as part of the `PrivateServer` mode. All API routes have been moved to `/api`, and the front-end `baseURL` has been adjusted accordingly.

### CODE REVIEW

File: server/bleep/src/webserver.rs [hunk 1]
```diff
@@ -65,33 +66,45 @@ pub async fn start(app: Application) -> Result<()> {
...
-        .with_state(app)
+        .with_state(app.clone())
         .layer(CorsLayer::permissive())
         .layer(CatchPanicLayer::new());
 
+    let mut router = Router::new().nest("/api", api);
+
+    if let Environment::PrivateServer = app.env {
+        router = router.merge(SpaRouter::new(
+            "/",
```
Comment (anastasiya1155) [inline server/bleep/src/webserver.rs:100]: we need more than the root route here, if the app is showing /search page it results in 404. We need all endpoints except for /api to serve FE

File: server/bleep/src/remotes/poll.rs [hunk 2]
```diff
@@ -26,7 +26,7 @@ const POLL_INTERVAL_MINUTE: &[Duration] = &[
 
 pub(crate) async fn check_credentials(app: Application) {
     loop {
-        if app.env.use_aaa() {
+        if let Environment::PrivateServer = app.env {
```
Comment (rsdy) [inline server/bleep/src/remotes/poll.rs:29]: Don't do direct matching of these. Check for individual permissions bound to the server mode, as it was originally, this helps us understand the different modes the app can run in easily.

File: server/bleep/src/webserver.rs [hunk 3]
```diff
@@ -65,33 +66,45 @@ pub async fn start(app: Application) -> Result<()> {
...
 
     if app.env.allow_path_scan() {
-        router = router.route("/repos/scan", get(repos::scan_local));
+        api = api.route("/repos/scan", get(repos::scan_local));
     }
 
     // Note: all routes above this point must be authenticated.
-    if app.env.use_aaa() {
-        router = aaa::router(router, app.clone());
+    if let Environment::PrivateServer = app.env {
```
Comment (rsdy) [inline server/bleep/src/webserver.rs:79]: Same here. Revert this change.

File: server/bleep/src/webserver.rs [hunk 4]
```diff
@@ -65,33 +66,45 @@ pub async fn start(app: Application) -> Result<()> {
...
         .layer(Extension(app.semantic.clone()))
         .layer(Extension(app.clone()))
-        .with_state(app)
+        .with_state(app.clone())
         .layer(CorsLayer::permissive())
         .layer(CatchPanicLayer::new());
 
+    let mut router = Router::new().nest("/api", api);
+
+    if let Environment::PrivateServer = app.env {
```
Comment (rsdy) [inline server/bleep/src/webserver.rs:98]: And here you should introduce a `app.env.serve_frontend()` boolean.

**Review by rsdy — CHANGES_REQUESTED** [review]

File: server/bleep/src/webserver.rs [hunk 5]
```diff
@@ -65,33 +66,45 @@ pub async fn start(app: Application) -> Result<()> {
...
+        router = router.merge(SpaRouter::new(
+            "/",
+            app.config
+                .frontend_dist
+                .as_ref()
+                .expect("need path to built frontend folder"),
```
Comment (rsdy) [inline server/bleep/src/webserver.rs:104]: I don't think this should be a hard error, just continue if not found. We could have an API running separately from the frontend assets due to eg. CDN deployments. A hard error makes this implementation dependent instead of configuration dependent.

**Comment by rsdy** [comment]
See #119 for packaging

File: server/bleep/src/lib.rs [hunk 6]
```diff
@@ -141,14 +141,12 @@ impl Environment {
...
-        match self {
-            Server => false,
-            InsecureLocal => false,
-            PrivateServer => true,
-        }
+        matches!(self, Self::PrivateServer)
+    }
+
+    pub(crate) fn serve_frontend(&self) -> bool {
+        matches!(self, Self::PrivateServer)
```
Comment (llogiq) [inline server/bleep/src/lib.rs:149]: Do we need two methods for the same thing here?

**Review by llogiq — APPROVED** [review]
Just a very small nit code-wise. Otherwise, some doc comments (what a private server is (for), how to set it up) would be helpful.

File: server/bleep/src/lib.rs (hunk 6 above)
Comment (calyptobai) [reply server/bleep/src/lib.rs:149]: @rsdy requested this, I'll revert the change
